src/ui/SearchInterface/ScrollRestoration.ts

```diff
@@ -0,0 +1,89 @@
+import { ResultListEvents, IDisplayedNewResultsEventArgs } from '../../events/ResultListEvents';
+import { $$ } from '../../utils/Dom';
+import { Logger } from '../../misc/Logger';
+import { StorageUtils } from '../../utils/StorageUtils';
+import { QueryStateModel } from '../../models/QueryStateModel';
+
+export interface IScrollRestorationInfo {
+  lastPosition: number;
+  pageHeight: number;
+}
+
+export class ScrollRestoration {
+  static ID = 'ScrollRestoration';
+
+  private scrollInfoStorage: StorageUtils<Record<string, IScrollRestorationInfo>>;
+
+  private restorationTimeOutInMs: number = 5000;
+  private tryToScrollIntervalInMs: number = 50;
+  private timeoutHandle = null;
+
+  constructor(public element: HTMLElement, private queryStateModel: QueryStateModel) {
+    this.scrollInfoStorage = new StorageUtils<Record<string, IScrollRestorationInfo>>(ScrollRestoration.ID, 'session');
+    window.addEventListener('beforeunload', () => {
+      this.saveScrollPositionInState();
+    });
+    $$(this.element).on(ResultListEvents.newResultsDisplayed, (event: Event, args: IDisplayedNewResultsEventArgs) =>
+      this.handleNewResultsDisplayed(args)
+    );
+  }
+
+  public saveScrollPositionInState() {
+    const scrollInfo: IScrollRestorationInfo = {
+      pageHeight: window.document.body.scrollHeight,
+      lastPosition: window.pageYOffset
+    };
+
+    const key = this.getKeyForCurrentQuery();
+    const prevScrollInfo = this.scrollInfoStorage.load();
+    this.scrollInfoStorage.save({
+      ...prevScrollInfo,
```

Comment:
Why do we need to "merge" with the old info ? Instead of overwrite completely ?

Reply:
We could have scroll info for another queries/pages in the same session. Imagine a user in a search interface with a certain query, then clicking in a result, then going to another search page with new parameters, and so on. I think makes sense to preserve scroll information of multiple possible search pages.

Reply:
👍 Makes sense !